File: core/src/main/java/iskake/jgbe/core/gb/mem/CartridgeROM.java
package iskake.jgbe.core.gb.mem;

import iskake.jgbe.core.gb.mem.mbc.MemoryBankController;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

/**
 * Read Only Memory of 'Game Boy game pak', separated into multiple
 * {@code ROMBank}s.
 */
public class CartridgeROM implements ReadableMemory, WritableMemory {
    private static final Logger log = LoggerFactory.getLogger(CartridgeROM.class);
    private final String romName;
    private final ROMBank[] ROMBanks;
    private final MemoryBankController mbc;
    private final RAM[] RAMBanks;

    CartridgeROM(String romName, ROMBank[] romBanks, RAM[] ramBanks, MemoryBankController mbc) {
        this.romName = romName;
        this.ROMBanks = romBanks;
        this.RAMBanks = ramBanks;
        this.mbc = mbc;
    }

    public void init() {
        mbc.init();
    }

    /**
     * Try to restore all RAM banks, if possible.
     * Will attempt to read the 'save RAM' file if the ROM has battery support.
     */
    public void tryRestoreRAM() {
        if (mbc.battery && (RAMBanks != null)) {
            try {
                String inName = getPathNoExt(romName);
                FileInputStream fp = new FileInputStream(inName);
                for (RAM bank : RAMBanks) {
                    fp.read(bank.bytes);
                }
                fp.close();
                log.info("Read saveram: " + inName);
            } catch (IOException ignored) {
                log.warn("Tried to restore save RAM, but failed.");
            }
        }
    }

    /**
     * Try to save all RAM banks, if possible.
     * Will attempt to write all RAM banks if the ROM has battery support.
     */
    public void trySaveRAM() {
        if (mbc.battery && (RAMBanks != null)) {
            try {
                String outName = getPathNoExt(romName);
                FileOutputStream fp = new FileOutputStream(outName);
                for (RAM bank : RAMBanks) {
                    fp.write(bank.bytes);
                }
                fp.flush();
                fp.close();
                log.info("Wrote saveram: " + outName);
            } catch (IOException ignored) {
                log.warn("Tried to write save RAM to file, but failed.");
            }
        }
    }

    private static String getPathNoExt(String path) {
        if (path.endsWith(".gb") || path.endsWith(".gbc")) {
            return path.substring(0, path.indexOf(".")) + ".sav";
        } else {
            log.warn("The provided file does not have the extension: `.gb` or `.gbc`. Appending `.sav` to filename...");
            return path + ".sav";
        }
    }

    /**
     * Get the ROM bank with index 0.
     *
     * @return The correct ROM bank.
     */
    public ROMBank getROMBank0() {
        // ? Some MBCs can switch the first bank to other banks than bank0.
        return ROMBanks[mbc.getROMBank0Index()];
    }

    /**
     * Get the current switchable ROM bank, according to the MBC.
     *
     * @return The correct ROM bank.
     */
    public ROMBank getROMBankX() {
        return ROMBanks[mbc.getROMBankNIndex()];
    }

    /**
     * Get the current RAM bank, if any.
     *
     * @return The current RAM bank. If there is none, then {@code null} is
     *         returned instead.
     */
    public RAM getRAMBank() {
        if (RAMBanks.length == 0 || !mbc.isRAMEnabled())
            return null;

        return RAMBanks[mbc.getRAMBankIndex()];
    }

    @Override
    public void write(int address, byte value) throws IndexOutOfBoundsException {
        mbc.write(address, value);
    }

    @Override
    public byte read(int address) throws IndexOutOfBoundsException {
        if (address < 0x4000) {
            return getROMBank0().read(address);
        } else {
            address -= 0x4000;
            return getROMBankX().read(address);
        }
    }
}

File: core/src/main/java/iskake/jgbe/core/gb/mem/CartridgeROM.kt
package iskake.jgbe.core.gb.mem

import iskake.jgbe.core.NotImplementedException
import iskake.jgbe.core.gb.ROMHeader

/**
 * Read Only Memory of 'Game Boy game pak', separated into multiple
 * `ROMBank`s.
 */
class CartridgeROM(bytes: ByteArray?) : ReadableMemory, WritableMemory {
    private val romBanks: Array<ROMBank?>
    private val mbc: MemoryBankController?
    private val ramBanks: Array<RAM?>

    init {
        var numROMBanks = ROMHeader.getROMBanksNum(bytes)
        if (numROMBanks == -1) {
            System.err.println("Invalid/unknown ROM bank size, assuming no extra banks...")
            numROMBanks = 2

            // ?Also possible fallback:
            // numROMBanks = (bytes.length / ROMBank.BANK_SIZE);
            // ?If this is used, should we try to assume the MBC type too?
        }
        romBanks = arrayOfNulls(numROMBanks)
        for (i in 0 until numROMBanks) {
            val offset = i * ROMBank.BANK_SIZE
            romBanks[i] = ROMBank((bytes!!).copyOfRange(offset, offset + ROMBank.BANK_SIZE))
        }
        val tmpMBC: MemoryBankController? = try {
            ROMHeader.getMBCType(romBanks[0])
        } catch (e: NotImplementedException) {
            System.err.println("Unimplemented/unknown MBC type, assuming no MBC...")
            NoMBC()
        } catch (e: IllegalArgumentException) {
            System.err.println("Unimplemented/unknown MBC type, assuming no MBC...")
            NoMBC()
        }
        mbc = tmpMBC
        var numRAMBanks = ROMHeader.getRAMBanksNum(romBanks[0])
        if (numRAMBanks == -1) {
            System.err.println("Invalid/unknown RAM bank size, assuming no external RAM...")
            numRAMBanks = 0
        }
        ramBanks = Array(numRAMBanks) { RAM(0x2000) }
    }// ? Some MBCs can switch the first bank to other banks than bank0.

    /**
     * Get the ROM bank with index 0.
     *
     * @return The correct ROM bank.
     */
    val rOMBank0: ROMBank?
        get() =// ? Some MBCs can switch the first bank to other banks than bank0.
            romBanks[0]

    /**
     * Get the current switchable ROM bank, according to the MBC.
     *
     * @return The correct ROM bank.
     */
    val romBankX: ROMBank?
        get() = romBanks[mbc!!.switchableIndex]// TODO: Test what reads to EXTRAM with no RAM bank returns.

    /**
     * Get the current RAM bank, if any.
     *
     * @return The current RAM bank. If there is none, then `null` is
     * returned instead.
     */
    val ramBank: RAM?
        get() = ramBanks[0] // TODO: Test what reads to EXTRAM with no RAM bank returns.

    @Throws(IndexOutOfBoundsException::class)
    override fun write(address: Int, value: Byte) {
        if (mbc == null) return
        mbc.write(address, value)
    }

    @Throws(IndexOutOfBoundsException::class)
    override fun read(address: Int): Byte {
        var addr = address
        return if (addr < 0x4000) {
            rOMBank0!!.read(addr)
        } else {
            addr -= 0x4000
            romBankX!!.read(addr)
        }
    }
}